Feature: Add Options for preview images in FileManager and bring back image popover on hover

haeflimi opened the pull request:
Back in concrete5.6 we had enlarged preview images in the FileManager when hovering over images. This was very useful in projects where we managed a lot of pictures in the FileManager. It allowed us to make a clear destinctions between two somewhat similar images in the FileManager listing without having to open the detail preview.
With the limited 30x30 thumbnails we have to work with since 5.7 this kind of image management became much more tedious since the extra step of opening the full image and closing it again, just to make sure we are about to select the right one, is necessary more often than not.

This PR adds a few options that allow us to fine tune the way preview images are displayed in the file manager:

`
'file_manager' => [

        'images' => [
            /*
             * The style of preview image used in the file_manager
             *
             * @var string 'small'(default,30x30), 'large(60x60)' or 'full(size of file_manager_listing)'
             */
            'preview_image_size' => 'small',
            /*
             * Show file_manager_detail thumbnail as preview image in popover
             *
             * @var boolean
             */
            'preview_image_popover' => false
        ],
    ],`

I didn't want to introduce any new default tumbnail types, so the options are still rather limited. However. 'preview_image_size' => 'full' allows for full flexibility to enlarge the preview image to whatever size we want by adjusting the thumbnail size of 'file_manager_listing'  in /dashboard/system/files/thumbnails while 'preview_image_popover' brings back the hover style preview that we know from 5.6 and earlier.

What do you guys think? - Is this useful/ flexible enough to concider for the core?

I was contemplating on if/ where to put these options in the dashboard. /dashboard/system/files/image_uploading seems to be the most fitting place even though most options there are concerned with processing of uploaded images rather than options for the FileManager. - What do you think?

MrKarlDilkington commented on concrete/js/build/core/file-manager/search.js:
```diff
@@ -373,6 +372,22 @@
+                animation: true,
+                html: true,
+                content: '<img class="img-responsive" src="'+image.src+'" alt="Thumbnail"/>',
+                container: '#ccm-dashboard-content',
```
What about switching the container and placement. See comment for screenshots.
```
container: '.ccm-file-manager-list',
placement: 'right',
```

MrKarlDilkington commented:
@haeflimi 

Thank you for adding this.

If we switch the popover container and placement, we can offer an image preview in the file manager modal. Also, it appears that placement `auto` is not dynamically adjusting the popover position, using `right` appears to work a little better. There is still a little clipping in the file manager modal because of overflow set to auto.
```
container: '.ccm-file-manager-list',
placement: 'right',
```

**Example:**

![image 001](https://user-images.githubusercontent.com/10898145/38963563-931f581e-4340-11e8-81d4-12d6abe70c5b.png)

![image 002](https://user-images.githubusercontent.com/10898145/38963564-933121ca-4340-11e8-8c99-cdba3f10fe99.png)

![image 003](https://user-images.githubusercontent.com/10898145/38963565-933f0768-4340-11e8-9209-55481ccecf9a.png)

![image 004](https://user-images.githubusercontent.com/10898145/38963562-930e01d6-4340-11e8-80b9-da941e04a433.png)

I think it would be very useful to put the file manager `preview_image_size` and `preview_image_popover` config settings in the dashboard somewhere. Even with the preview image popover, I imagine many people will still want a larger default thumbnail, so making the setting easily accessible is important. I am not sure where to put the settings either.

haeflimi commented:
@MrKarlDilkington and @Mesuva  Thanks for your feedback.

I've actually tried different configurations with **placement** and container attachment and found that auto was the only configuration that worked reliably even if it does appear a bit laggy when opening the popover. - I agree that the placement to the right feels the most intuitive and looks best. But i had a problem with the popover placement that i couldn't get around. I am suprised that you didn't encounter this because you 've obviously tested it too. Here's what happens on my dev. environment when i set the placement/ container as you suggest.:

When no images are cached in the Browser and the thumbnail is hovered for the first time the popover can't be placed correctly and appears as if an entry several lines down is hovered.:
https://www.useloom.com/share/6cf2dd48fcad496eb5dd5961e704b19f
On subsequent hovers it works correctly as expected.

I think the reason for that is the popover beeing initialized before the image is loaded and it's dimensions are known so the image can't be vertically centered in relation to the position where it should appear. Placing the popover auto (which seems to make it top or bottom depending on space available.) works around this problem by not having to account for the images hieght when placing the popover in the vertical axis.

The differnet container attachment @MrKarlDilkington suggested works better though, so i've addopted that. And while fiddling around a little more i've also removed the popover's width restriction so the images in the popover appear in the the dimensions that are defined for 'file_manager_detail'

If someone happens to have an idea on how to get around the vertical placement problem and make placement: right work properly. I'm all ears.

MrKarlDilkington commented:
@haeflimi 

I spent a while last night trying to find ways to add the popovers to the right, and each attempt failed in some way. Ultimately, to get them to position correctly to the right, you need to set an explicit height on the img element. The problem with that is at narrow screen widths, the images are not responsive.

Regarding where to add the config options in the dashboard, I think they fit on the Accessibility page.

**Example:**

![image](https://user-images.githubusercontent.com/10898145/39080059-271fd254-44f5-11e8-9e22-e001f8f46110.png)

MrKarlDilkington commented:
@aembler 

Are you open to having preview popups on by default, but also including the two Accessibility page options to turn the preview popups off and increase preview image size (large preview off by default)?

I've heard comments from some community members that they would prefer larger preview images. This offers something for everyone without affecting the default state.

haeflimi commented:
I've Added the Options to the Dashboard as @MrKarlDilkington suggested. - But changed the default for the image popover to true as @aembler did.
I don't think we should force the popover upon everyone. - With changes like this, that in the end boil down to individual preference i think it's good to have the option to turn it off.

The clipping Problem in the Modal situation should be fixed now. For that i needed to change the overflow behaviour on the File Manager Dialog. - I hope by doing that i didn't break anything else. As far as i've tested it looks ok.

On the topic of mobile: Since the hover state isn't really a thing on mobile/ touch devices i don't think we have to do anything speciffically. - I might be wrong though.

MrKarlDilkington commented:
@haeflimi 

Removing `overflow: visible;` on div.ui-dialog .ui-dialog-content breaks modal scrolling and does not prevent popover preview clipping.

![image](https://user-images.githubusercontent.com/10898145/39651089-907ed29a-4fb8-11e8-94fe-aeb036d0d8ab.png)

To prevent clipping, I believe we need to set the `container` property to an element that is "outside" or a parent of the modal. The element would need to be shared on front-end pages and dashboard pages.

haeflimi commented:
@MrKarlDilkington 
You are right. - Would have been to good to be true if  that worked  without breaking anything.

> To prevent clipping, I believe we need to set the container property to an element that is "outside" or a parent of the modal. The element would need to be shared on front-end pages and dashboard pages.

I've tried that before. - The problem there was: At some point we lost the Styles for the popover element since all BS styles are namespaced.

Now i've undone the overflow change, Attached the popover to body and added/ included some extra css to style the popover and adjust the z-Index.
Now it seems to be working/ looking smoothly and without clipping.

MrKarlDilkington commented:
@haeflimi 

The popovers appear to "stutter" when you hover them on first load. They appear in one position, then switch to another, please see video.
https://www.useloom.com/share/e5a19ad9d8a14c3ca31f2476118868b1

haeflimi commented:
@MrKarlDilkington 
I see what you mean. On my local dev Environment this didn't really show since Images are loaded much quicker which makes the stutter bearly noticable.
The issue here is the auto placement for the popover. - it can only be positioned in the final position when the image is completely loaded. In the same way as the position: right method is unable to correctly position it in the vertical axis.

I think i've solved that now, which also makes placement: right work properly too.
I needed to deactivate the default popover hover trigger behaviour, load the image manually on hover and only trigger the popover when the image is completely loaded.

MrKarlDilkington commented:
@haeflimi 

I tested the latest changes and everything looks to be working as expected.

aembler commented:
This is really nice, and it's something that's long overdue. This was a nice feature in 5.6 and I'd like to see it come back. You've done a lot of good work getting it closer.

I do have some concerns about what I'm seeing here. I'll just post them in no particular order:

1. We need to limit the size of the thumbnail somehow. Look what it looks like when I mouse over one of the image slider images:

<img width="1204" alt="screen shot 2018-05-23 at 1 49 12 pm" src="https://user-images.githubusercontent.com/527809/40450351-41fac354-5e90-11e8-9975-793ccf91cf7b.png">

2. I don't like how the popover arrow is actually hovering further down next to an image that isn't the one that I'm hovering over.
3. I think it'd be better if the image hovered below the item – right now if I accidentally hover over the thumbnail it obscures the metadata about the image i'm interested in.
4. I'm not sure we need all of these configuration settings. I think if we can get this working smoothly we can just default it to on and be done with it. Although I'm flexible on this last point.

aembler commented:
Again, this is a really great start, and it looks like we're really close.

haeflimi commented:
> Again, this is a really great start, and it looks like we're really close.

Don't worry i haven't abandoned it. - I am still on it (when i find the time).

@aembler 
I think you might not had the most recent version of the .js code after my last edit to the PR. (maybe because of the confilict on the file-manager.js. I dont really know how i am supposed to merge the grunt generated .js in case of a conflict - sorry about that.)

To address your concernes:

1. - I've added a max-width of 75vw to the image in the popover. - This shoud not only improve things in the situation of a overly large image but also icrease usability on small screens.

2. - This is related to the Clipping issue i had and should be resolved with my last edit. - The image/ arrow should appear in the right place. @MrKarlDilkington has tested/ alredy confirmed this too.

3. - I haven't thought about the ability to see the metadata next to the image and it's a good point. - I've now changed the popover to use bottom placement. - There is a downside though: If the last image in the list is a large one. - It might not fit into the space between the bottom line of the File Entry and the end of the list and thuss can't be seen fully. - I don't see a easy way to have both (guaranteed visibility on the metadata using bottom placement and guaranteed full visibility of the preview image.)

4. - I'd personally like to have the configuration options for different projects/ needs. So i'd rather not remove them.

aembler commented:
I don't think I'm seeing the latest changes. I deleted my branch for this and pulled again from your branch, and merged the changes in, and it doesn't look any different to me. The arrow isn't in the right place, the window isn't limited in width. Is there something that may need to get added to the pull request?

![screen shot 2018-05-30 at 3 26 25 pm](https://user-images.githubusercontent.com/527809/40750932-f5d4998a-641d-11e8-8848-ad13ecb69068.png)

hissy commented:
Our client says file manager listing thumbnail is too small. I agree also. Let's merge this feature!

aembler commented:
@hissy I think my requests need to be resolved before we can merge this in. See comment https://github.com/concrete5/concrete5/pull/6593#issuecomment-391492269

I pulled in updates but I don't think they actually fixed the problems. I'd be happy to be proven wrong though, or if someone can go through and make those corrects we could definitely get this into 8.5.

haeflimi commented:
@hissy and @aembler 
I dont really see what there is left for me to fix. - I've just re- tested this with the latest develop branch + the PR and it works nicely for me. As i understood it @MrKarlDilkington did some testing as well and it seemd to work for him too.

As i've stated in Mai: The only thing that i am not sure about is if the uglified file-manager.js merged correctly in this PR. This can be easily resolved however by running the grunt build script after merging. Something i'd assuem would be done anyway when building a new c5 release.

I've updated the PR with a freshly rebuilt Version of the File in question. But since this didn't seem to resolve your issues the last time around, i dont think it makes a difference.

aembler commented:
Not sure what the issue was before, but I just merged this and I'm not seeing the same issues. Looks great! Thank you.

Seanom commented:
Is this functionality hidden behind a flag? I'm not seeing a preview image when hovering over images, I've upgraded an older 8.4 site to 8.5.1.
